File: Extra/155-Min-Stack.py
'''
Author: Ali Çine
Description:
    For the 155.Question we dont create a new data structure, we only implement "getMin" function to appropriate "list" data structure.To calculate the minimum element at O(1) time complexity, we must return a variable that is ready in memory. We check in every push if minimum element has changed and update minstack. We use list for minstack because of pop function, if we didnt use a list then we wouldnt get previous minimum if new minimum was popped.
'''
class MinStack:
    def __init__(self):
        self.stack = []
        self.minstack = []

    def push(self, val:int) -> None:
        minnumber = self.getMin()
        if minnumber is not None:
            if val < minnumber:
                self.minstack.append(val)
            else:
                self.minstack.append(minnumber)
        else:
            self.minstack.append(val)
        self.stack.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.minstack.pop()

    def top(self) -> int:
        return self.stack[-1] if self.stack else None

    def getMin(self) -> int:
        return self.minstack[-1] if self.stack else None

File: 2025-10-22/155-Min-Stack.cpp
//
// Created by egoistpizza on 18.10.2025.
//

// 155. Min Stack
// https://leetcode.com/problems/min-stack/
// * Amazon Interview Question *

// Time: O(1) for all operations
// Memory: O(n)

class MinStack {
private:

    stack<int> stack;
    stack<int> minStack;

public:

    MinStack() {}

    void push(int val) {
        stack.push(val);
        val = min(val, minStack.empty() ? val : minStack.top());
        minStack.push(val);
    }

    void pop() {
        stack.pop();
        minStack.pop();
    }

    int top() {
        return stack.top();
    }

    int getMin() {
        return minStack.top();
    }
};
